src/java/org/apache/cassandra/service/reads/repair/BlockingReadRepair.java

```diff
@@ -63,194 +55,168 @@
+        public DigestRepair(DataResolver dataResolver, ReadCallback readCallback, Consumer<PartitionIterator> resultConsumer, ReplicaCollection initialContacts)
         {
             this.dataResolver = dataResolver;
             this.readCallback = readCallback;
             this.resultConsumer = resultConsumer;
+            this.initialContacts = initialContacts;
         }
     }
 
     public BlockingReadRepair(ReadCommand command,
+                              // TODO: remove
                               ReplicaList replicas,
                               long queryStartNanoTime,
                               ConsistencyLevel consistency)
     {
         Replicas.checkFull(replicas);
         this.command = command;
-        this.replicas = replicas;
         this.queryStartNanoTime = queryStartNanoTime;
         this.consistency = consistency;
+        this.cfs = Keyspace.openAndGetStore(command.metadata());
     }
 
     public UnfilteredPartitionIterators.MergeListener getMergeListener(ReplicaList replicas)
     {
-        return new PartitionIteratorMergeListener(replicas, command, this);
+        return new PartitionIteratorMergeListener(replicas, command, consistency, this);
     }
 
-    public static class BlockingPartitionRepair extends AbstractFuture<Object> implements RepairListener.PartitionRepair
+    /**
+     * Blocking repairs can be started by a digest mismatch for single partition reads, merged data
+     * on a partition range read, or transient data being repaired to full replicas. This makes sure
+     * we only report one blocking repair for each read repair instance.
+     */
+    private void maybeMarkBlockingRepair()
     {
-
-        final List<AsyncOneResponse<?>> responses;
-        final ReadCommand command;
-        final ConsistencyLevel consistency;
-
-        public BlockingPartitionRepair(int expectedResponses, ReadCommand command, ConsistencyLevel consistency)
+        if (!blockingRepairReported)
         {
-            this.responses = new ArrayList<>(expectedResponses);
-            this.command = command;
-            this.consistency = consistency;
+            ReadRepairMetrics.repairedBlocking.mark();
+            blockingRepairReported = true;
         }
+    }
 
-        private AsyncOneResponse sendRepairMutation(Mutation mutation, InetAddressAndPort destination)
+    public void startRepair(DigestResolver digestResolver, ReplicaList allReplicas, ReplicaList contactedReplicas, Consumer<PartitionIterator> resultConsumer)
+    {
+        maybeMarkBlockingRepair();
+
+        // Do a full data read to resolve the correct response (and repair node that need be)
+        Keyspace keyspace = Keyspace.open(command.metadata().keyspace);
+        DataResolver resolver = new DataResolver(keyspace, command, ConsistencyLevel.ALL, allReplicas, this, allReplicas.size(), queryStartNanoTime);
```

Comment:
Huh we repair at CL.ALL, but there is plumbing to speculate with additional replicas?